### PR TITLE
Always initialize MMF + Implement new LUA functions

### PR DESCRIPTION
**Pull request by E-Sh4rk**
This commit adresses two things:

1. Currently, the MMF LUA functions are only available if Bizhawk is started with the "--mmf=filename" flag.
I think this is not intuitive, because we should not have to specify a filename in order to active MMF functions
as the given "filename" is only used for "mmfScreenshot" (and even for this function, the filename can be set in the LUA).
I changed this behavior so that MMF functions are always available, and "mmfScreenshot" will fail with a log if
no filename has been specified by the command line nor by the LUA script.

2. I have added several LUA functions:

- "memory.writebyterangeataddress": This is because the function "memory.writebyterange" does not match with what "memory.readbyterange" returns...
"memory.readbyterange" returns a table with the keys going from 0 to len-1, while "memory.writebyterange" returns a table with each key being the full address (not starting at 0).
Actually, "memory.writebyterange" is not only for writing a range, by any set of addresses.
On the other hand, the new function "memory.writebyterangeataddress" has a behavior similar to "memory.readbyterange" or "comm.socketServerSendBytes":
the lua table given as argument is seen as an array instead of an (address, value) dictionnary.

- "mmfWriteBytes", "mmfReadBytes": to send/read a sequence of bytes to/from a memory mapped file

- "mmfCopyToMemory", "mmfCopyFromMemory": to copy the content of a memory mapped file to a given location in the emulator memory (and the other direction).
Note that this could be achieved directly in the LUA script with the combination of "memory.readbyterange" and "mmfWriteBytes", but is is A LOT slower
(probably because of the type conversions). Also note that I personaly need these functions in order to be able to efficiently read/write to the game memory
from an external tool.

### CODE REVIEW

File: src/BizHawk.Client.Common/Api/MemoryMappedFiles.cs [hunk 1]
```diff
@@ -8,33 +8,44 @@ namespace BizHawk.Client.Common
...
+			{
+				mmfFile = _mmfFiles[filename] = MemoryMappedFile.OpenExisting(filename);
+			}
+
+			var viewAccessor = mmfFile.CreateViewAccessor(0, expectedSize, MemoryMappedFileAccess.Read);
 			var bytes = new byte[expectedSize];
 			viewAccessor.ReadArray(0, bytes, 0, expectedSize);
-			return Encoding.UTF8.GetString(bytes);
+			viewAccessor.Dispose();
+			return bytes;
```
Comment (YoshiRulz) [inline src/BizHawk.Client.Common/Api/MemoryMappedFiles.cs:37]: Why replace `using` with `Dispose` here?

File: src/BizHawk.Client.Common/Api/MemoryMappedFiles.cs [hunk 2]
```diff
@@ -8,33 +8,44 @@ namespace BizHawk.Client.Common
...
 		}
 
-		public int ScreenShotToFile() => WriteToFile(Filename, _takeScreenshotCallback());
+		public int ScreenShotToFile() => WriteToFile(Filename!, _takeScreenshotCallback());
```
Comment (YoshiRulz) [inline src/BizHawk.Client.Common/Api/MemoryMappedFiles.cs:40]: This should probably have better error handling:
```suggestion
		public int ScreenShotToFile()
		{
			if (Filename is null) {
				Console.WriteLine("MMF screenshot target not set; start EmuHawk with `--mmf=filename`");
				return 0;
			}
			return WriteToFile(Filename, _takeScreenshotCallback());
		}
```

And then the filename doesn't need to be checked in `CommLuaLibrary.MmfScreenshot`.

File: src/BizHawk.Client.Common/Api/MemoryMappedFiles.cs [hunk 3]
```diff
@@ -8,33 +8,44 @@ namespace BizHawk.Client.Common
...
 
 		public int WriteToFile(string filename, byte[] outputBytes)
 		{
 			int TryWrite(MemoryMappedFile m)
 			{
 				using var accessor = m.CreateViewAccessor(0, outputBytes.Length, MemoryMappedFileAccess.Write);
 				accessor.WriteArray(0, outputBytes, 0, outputBytes.Length);
+				accessor.Dispose();
```
Comment (YoshiRulz) [inline src/BizHawk.Client.Common/Api/MemoryMappedFiles.cs:48]: This is superfluous, the local is declared with `using` already.

File: src/BizHawk.Client.Common/lua/CommonLibs/CommLuaLibrary.cs [hunk 4]
```diff
@@ -133,43 +133,55 @@ private bool CheckSocketServer()
...
+		}
 		[LuaMethod("mmfRead", "Reads a string from a memory mapped file")]
 		public string MmfRead(string mmf_filename, int expectedSize)
 		{
-			CheckMmf();
-			return APIs.Comm.MMF?.ReadFromFile(mmf_filename, expectedSize);
+			return APIs.Comm.MMF.ReadFromFile(mmf_filename, expectedSize);
 		}
-
-		private void CheckMmf()
```
Comment (YoshiRulz) [inline src/BizHawk.Client.Common/lua/CommonLibs/CommLuaLibrary.cs:167]: This is hard to read, can you split the changeset into two commits? One to make `--mmf` optional, and one to add extra functions.

File: src/BizHawk.Client.Common/lua/CommonLibs/MemoryLuaLibrary.cs [hunk 5]
```diff
@@ -47,12 +47,23 @@ public MemoryLuaLibrary(IPlatformLuaLibEnv luaLibsImpl, ApiContainer apiContaine
...
+		[LuaMethod("writebyterangeataddress", "Writes the given values as a contiguous block of unsigned bytes to a specific address")]
+		public void WriteByteRangeAtAddress(LuaTable memoryblock, long addr, string domain = null)
+		{
+			int i = 0;
+			foreach (var v in _th.EnumerateValues<double>(memoryblock))
+			{
+				APIs.Memory.WriteByte(addr+i, (uint)v, domain);
+				i++;
+			}
+		}
```
Comment (YoshiRulz) [inline src/BizHawk.Client.Common/lua/CommonLibs/MemoryLuaLibrary.cs:63]: This is out of scope, and we haven't reached a consensus re: #2844 yet.

**Review by YoshiRulz — CHANGES_REQUESTED** [review]
I have no problems with this conceptually. The MMF stuff is niche enough that you can add whatever you need.

File: src/BizHawk.Client.Common/lua/CommonLibs/MemoryLuaLibrary.cs (hunk 5 above)
Comment (E-Sh4rk) [reply src/BizHawk.Client.Common/lua/CommonLibs/MemoryLuaLibrary.cs:63]: Ok, I am gonna remove this function!

File: src/BizHawk.Client.Common/lua/CommonLibs/CommLuaLibrary.cs (hunk 4 above)
Comment (E-Sh4rk) [reply src/BizHawk.Client.Common/lua/CommonLibs/CommLuaLibrary.cs:167]: Sure. I dont know if I can do that in this pull request, so maybe I will close this one and open a new one.

Comment (vadosnaprimer) [reply src/BizHawk.Client.Common/lua/CommonLibs/CommLuaLibrary.cs:167]: You can force-push whatever set of commits you consider proper.

**Comment by E-Sh4rk** [comment]
I'll make a new pull request with the changes you requested.
